ui/packages/catalog-ui-search/src/main/webapp/js/controllers/cesium.layerCollection.controller.js

```diff
@@ -43,96 +114,127 @@ var Controller = CommonLayerController.extend({
+            : DEFAULT_HTTP_PORT
+      }
+      Cesium.TrustedServers.add(parsedUrl.hostname, port)
+    }
+
+    this.layerOrder = LayerOrdering.addLayer({
+      prev: this.layerOrder,
+      cur: this.collection.models.map(model => model.id).reverse(),
+      layer: model.id,
+    })
+
+    const layerIndex = this.layerOrder.indexOf(model.id)
+    const layer = this.map.imageryLayers.addImageryProvider(
+      provider,
+      layerIndex
+    )
+
+    this.layerForCid[model.id] = layer
+    layer.alpha = model.get('alpha')
+    layer.show = model.shouldShowLayer()
+  },
   onDestroy: function() {
     if (this.isMapCreated) {
       this.map.destroy()
       this.map = null
     }
   },
   setAlpha: function(model) {
-    var layer = this.layerForCid[model.id]
+    const layer = this.layerForCid[model.id]
     layer.alpha = model.get('alpha')
   },
   setShow: function(model) {
-    var layer = this.layerForCid[model.id]
+    if (!this.layerForCid[model.id]) {
+      this.initLayer(model)
+    }
+    const layer = this.layerForCid[model.id]
     layer.show = model.shouldShowLayer()
   },
   /*
-            removing/re-adding the layers causes visible "re-render" of entire map;
-            raising/lowering is smoother.
-            raising means to move to a higher index.  higher indexes are displayed on top of lower indexes.
-            so we have to reverse the order property here to make it display correctly.  
-            in other words, order 1 means highest index.
-        */
+    removing/re-adding the layers causes visible "re-render" of entire map;
+    raising/lowering is smoother.
+    raising means to move to a higher index.  higher indexes are displayed on top of lower indexes.
+    so we have to reverse the order property here to make it display correctly.  
+    in other words, order 1 means highest index.
+  */
   reIndexLayers: function() {
-    this.collection.forEach(function(model, index) {
-      var layer = this.layerForCid[model.id]
-      var previousOrder = this.map.imageryLayers.indexOf(layer) + 1
-      var currentOrder = this.collection.length - model.get('order') // order is backwards on cesium (higher indexes are displayed above lower)
-      var method = currentOrder > previousOrder ? 'raise' : 'lower' // raise means move to higher index :(
-      var count = Math.abs(currentOrder - previousOrder)
-      // console.log(method + " " + model.get('name') + " " + count);  // useful for debugging!
-      _.times(
-        count,
-        function() {
-          this.map.imageryLayers[method](layer)
-        },
-        this
-      )
-    }, this)
+    const newLayerOrder = LayerOrdering.shiftLayers({
+      prev: this.layerOrder,
+      cur: this.collection.models.map(model => model.id).reverse(),
+    })
+    const { layer, method, count } = LayerOrdering.getShift({
+      prev: this.layerOrder,
+      cur: newLayerOrder,
+    })
+
+    if (typeof layer === 'undefined') {
+      return
```

Comment:
❓ Is this a common case? If not, should we log or throw an error?

Reply:
I think I might remove this and throw an error within the getShift function